File: flutter/lib/features/champion/data/datasource/patch_version_remote_data_source.dart
import 'package:dio/dio.dart';

class PatchVersionRemoteDataSource {
  const PatchVersionRemoteDataSource(this._dio);

  final Dio _dio;

  Future<String?> getLatestPatchVersion() async {
    final response = await _dio.get<List<dynamic>>('/api/versions.json');
    final versions = response.data;
    if (versions == null || versions.isEmpty) {
      return null;
    }

    final version = versions.first;
    return version is String && version.trim().isNotEmpty ? version : null;
  }
}

File: react/src/features/champion/data/data-sources/patch-version-remote-data-source.ts
import type { AxiosInstance } from 'axios';
import { z } from 'zod';

const patchVersionsSchema = z.array(z.string()).min(1);

export interface PatchVersionRemoteDataSource {
  fetchLatestVersion(): Promise<string>;
}

export class DataDragonPatchVersionRemoteDataSource implements PatchVersionRemoteDataSource {
  constructor(private readonly client: AxiosInstance) {}

  async fetchLatestVersion(): Promise<string> {
    const response = await this.client.get<unknown>('/api/versions.json');
    const latestVersion = patchVersionsSchema.parse(response.data)[0];
    if (!latestVersion) {
      throw new Error('Data Dragon returned an empty patch version list');
    }
    return latestVersion;
  }
}
